### ISSUE TITLE
Humidity range & tolerance

### ISSUE THREAD

**Issue by TeCoEd** [astro-pi/beta-testing#8]
Edit: Updated this issue to be about the range of the humidity slider and the tolerance of the sensor.  Original text below.

---

The sliders can be dragged over each other

![overlay](https://cloud.githubusercontent.com/assets/7890126/16817774/4f8a5a18-493c-11e6-86a4-2d187ded6941.jpg)

**Comment by eah13** [comment]
Thanks for reporting.  

I'm unable to reproduce in ChromeOS so could be an OS/browser specific issue.  If anyone else can repro this, please post here.  It'll be a day or two before I have access to a Windows machine to test.

**Comment by ebertmi** [comment]
@TeCoEd Can you please post your browser version

**Comment by TeCoEd** [comment]
Hi, I was using Chrome on Win10

**Comment by ebertmi** [comment]
@TeCoEd Can you please provide the full Chrome Version. You can find the version in the menu -> Help -> About Google Chrome

**Comment by eah13** [comment]
Unable to reproduce with Chrome 51.0.2704.106 on Windows.  @TeCoEd are you still seeing this, even in a new _Incognito_ window?  Incognito will start with a fresh cache, you're logged out of Trinket and most extensions are turned off.  If not, one of those things may be involved in the bug. If you still see it in Incognito, perhaps you could note any other details about your particular setup?  For instance, I tried this both logged in and logged out just in case that was a factor.

**Comment by ukscone** [comment]
can't reproduce on chrome on win10 or edge on win10 however the humidity slider goes to 110%

**Comment by eah13** [comment]
Good! The humidity slider's range is based on the sensor's range and 110 is a possible output.  I'm going to close this. We can reopen when someone reproduces.

**Comment by waveform80** [comment]
The sensor's operating range is 0 to 100 according to [the datasheet](http://www.st.com/content/ccc/resource/technical/document/datasheet/4d/9a/9c/ad/25/07/42/34/DM00116291.pdf/files/DM00116291.pdf/jcr:content/translations/en.DM00116291.pdf) (page 6) although one could infer a potential output of 105 given the accuracy limits of the outer range. As for the actual outputs from the device, the result will depend on the values in the calibration registers (page 25) which may vary from device to device. 110 is certainly possible, but given bad calibration one could produce almost any result (as the result depends on the slope between two points defined in the calibration registers).

It might be preferable to have the user set a value from 0 to 100 (the operating range) and have the simulation output values that vary around that setting (by an amount defined by the error ranges specified in the data-sheet, perhaps with some reasonable distribution like a normal variate)?

**Comment by ebertmi** [comment]
We can apply the error ranges (3.5% and 5%) when the user is reading the value. So the set humidity is between 0 and 100 and the simulation returns this value and applys the error ranges.

**Comment by waveform80** [comment]
> We can apply the error ranges (3.5% and 5%) when the user is reading the value. So the set humidity is between 0 and 100 and the simulation returns this value and applys the error ranges.

That sounds good. I'm assuming (sorry, haven't checked!) that trinket doesn't implement the `multiprocessing` or `threading` modules so there's no need to worry about simultaneous readers getting different readings - in which case applying the potential error on read is indeed a nice, simple means of simulating the HAT.

**Comment by ebertmi** [comment]
Correct, as trinket/skulpt is JavaScript based and JavaScript is basically single threaded (except workers, but those require a tremendous amount of work to implement) there will be never any problems.

**Comment by eah13** [comment]
@waveform80 thanks for the feedback.  I've reopened and update this issue to reflect the range/error issues we've identified.

**Comment by eah13** [comment]
Errors are now implemented!  You can test with this program:

```
from sense_hat import SenseHat
import time

sense = SenseHat()

green = (0, 255, 0)
white = (255, 255, 255)

while True:
    humidity = sense.humidity
    print(humidity)
    humidity_value = 64 * humidity / 100

    pixels = [green if i < humidity_value else white for i in range(64)]

    sense.set_pixels(pixels)
```

The printed value should be a gausian distribution based on the spec'd error ranges, which makes them 'jumpy' like the real Sense HAT readings. Thanks to @waveform80 for the idea and @waywaaard for the maths.

The other sensors got a similar treatment.  The error is applied in response to the sliders being changed instead of continually.

An update of the max range of the slider is forthcoming... will leave this issue open until then.

**Comment by eah13** [comment]
Humidity silder max is now 100%

**Comment by davidhoness** [comment]
This appears to not be fixed. This afternoon I gave a demonstration at the Raspberry Pi / Code Club all staff meeting and several people reported the sliders issue.

Screen shot from  Chrome Version 51.0.2704.103 (64-bit) on OS X Yosemite version 10.10

![screen shot 2016-07-20 at 14 45 40](https://cloud.githubusercontent.com/assets/6640424/16992019/9e9bfaea-4e95-11e6-9e12-a0f60eb9a183.png)

Happy if you want to leave this issue closed but raise a new one for the slider problem.

**Comment by brianpmarks** [comment]
@davidhoness Thanks for the update. Do you recall when the slider appeared this way? Did it initially load this way or perhaps after the bar between the code on the left and the emulator on the right was moved? Or something else entirely?

**Comment by davidhoness** [comment]
I believe it appeared like this after a page load.

**Comment by eah13** [comment]
OK fixed!  It took all of our brains combined but we were able to prevent the slider size bug in all cases we could replicate.  Extensive testing welcome.  Replication recipe we discovered:
- Create an error in a trinket and run it (`import sense_ha`)
- fix and run the code _without first dismissing the error bar_ (`import sense_hat`)

Recipe should look normal, but used to produce this bug.

For posterity, this bug had to do with resizing the screen elements after refreshing the Ace editor.  The sensor container elements had incorrectly calculated widths at various points and so were getting set to be huge.  We now wait til the elements have a valid width to update the rangesliders and refresh them at various points just to be sure.
